### PR TITLE
Fix BLE characteristics for notify and indicate.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: lib/ble.toit [hunk 1]
```diff
@@ -441,6 +441,23 @@ class LocalService extends Resource_ implements Attribute:
...
+        | CHARACTERISTIC_PERMISSION_READ_ENCRYPTED
+    read_properties_bits := CHARACTERISTIC_PROPERTY_READ
+        | CHARACTERISTIC_PROPERTY_NOTIFY
+        | CHARACTERISTIC_PROPERTY_INDICATE
+    write_permission_bits := CHARACTERISTIC_PERMISSION_WRITE
+        | CHARACTERISTIC_PERMISSION_WRITE_ENCRYPTED
+    write_properties_bits := CHARACTERISTIC_PROPERTY_WRITE
+        | CHARACTERISTIC_PROPERTY_WRITE_WITHOUT_RESPONSE
+
+    if permissions & read_permission_bits  != 0 and
```
Comment (kasperl) [inline lib/ble.toit:454]:
```suggestion
    if permissions & read_permission_bits != 0 and
```

File: src/resources/ble_esp32.cc [hunk 2]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
   ARGS(BleServiceResource, service_resource, Blob, raw_uuid, int, properties,
        int, permissions, Object, value, int, read_timeout_ms)
 
-  if (!service_resource->peripheral_manager()) INVALID_ARGUMENT;
+  if (!service_resource->peripheral_manager()) {
+    printf("service_resource->peripheral_manager() is null\n");
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:1980]: Drop the debug printing again?

File: src/resources/ble_esp32.cc [hunk 3]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
+    INVALID_ARGUMENT;
+  }
 
   ByteArray* proxy = process->object_heap()->allocate_proxy();
   if (proxy == null) ALLOCATION_FAILED;
 
-  if (service_resource->deployed()) INVALID_ARGUMENT;
+  if (service_resource->deployed()) {
+    printf("service_resource->deployed() is true\n");
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:1988]: Drop the debug printing again?

**Review by kasperl — COMMENTED** [review]
Didn't fully try to grok this yet.

File: src/resources/ble_esp32.cc [hunk 4]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
   }
 
   if (permissions & 0x2) { // WRITE
-    if (!(properties & (BLE_GATT_CHR_F_WRITE | BLE_GATT_CHR_F_WRITE_NO_RSP))) INVALID_ARGUMENT;
+    if (!(properties & (BLE_GATT_CHR_F_WRITE | BLE_GATT_CHR_F_WRITE_NO_RSP)))  {
+      INVALID_ARGUMENT;
+    }
   }
 
   if (permissions & 0x4) {
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:2007]:
```suggestion
  if (permissions & 0x4) {  // READ_ENCRYPTED.
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/ble.toit (hunk 1 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: src/resources/ble_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc [hunk 5]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
   if (proxy == null) ALLOCATION_FAILED;
 
-  if (service_resource->deployed()) INVALID_ARGUMENT;
+  if (service_resource->deployed()) {
+    printf("service_resource->deployed() is true\n");
+    INVALID_ARGUMENT;
+  }
 
   uint32 flags = properties & 0x7F;
   if (permissions & 0x1) {  // READ
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc [hunk 6]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
-    if (!(properties & BLE_GATT_CHR_F_READ)) INVALID_ARGUMENT;
+    if (!((properties & BLE_GATT_CHR_F_READ) ||
+          (properties & BLE_GATT_CHR_F_NOTIFY) ||
+          (properties & BLE_GATT_CHR_F_INDICATE))) {
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc [hunk 7]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
+      INVALID_ARGUMENT;
+    }
   }
 
   if (permissions & 0x2) { // WRITE
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 4 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc [hunk 8]
```diff
@@ -1976,29 +1976,47 @@ PRIMITIVE(add_characteristic) {
...
-    if (!(properties & BLE_GATT_CHR_F_READ)) INVALID_ARGUMENT;
+    if (!((properties & BLE_GATT_CHR_F_READ) ||
+          (properties & BLE_GATT_CHR_F_NOTIFY) ||
+          (properties & BLE_GATT_CHR_F_INDICATE))) {
+      INVALID_ARGUMENT;
+    }
     flags |= BLE_GATT_CHR_F_READ_ENC;  // _ENC = Encrypted.
   }
 
   if (permissions & 0x8) {
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

**Pull request merged**